In [1]:
import pandas as pd
import numpy as np 

In [2]:
books = pd.read_csv('book recommendation dataset/books.csv')
users = pd.read_csv('book recommendation dataset/users.csv')
ratings = pd.read_csv('book recommendation dataset/ratings.csv')

C:\Users\DELL\AppData\Local\Temp\ipykernel_3096\2343088646.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('book recommendation dataset/books.csv')


In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
books.shape, users.shape, ratings.shape

((271360, 8), (278858, 3), (1149780, 3))

## EDA 

In [7]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [8]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [9]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [10]:
books.duplicated().sum()

0

In [11]:
users.duplicated().sum()

0

In [12]:
ratings.duplicated().sum()

0

In [13]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [14]:
books['Publisher'].unique()

array(['Oxford University Press', 'HarperFlamingo Canada',
       'HarperPerennial', ..., 'Tempo', 'Life Works Books', 'Connaught'],
      dtype=object)

##### Top 10 Publishers

In [15]:
books.groupby('Publisher').size().reset_index().sort_values(by=0, ascending=False)[:10]

Publisher     0
6583                  Harlequin  7535
13667                Silhouette  4220
11762                    Pocket  3905
1391           Ballantine Books  3783
1421               Bantam Books  3646
13259                Scholastic  3160
13712      Simon &amp; Schuster  2971
11390             Penguin Books  2844
1717   Berkley Publishing Group  2771
15931              Warner Books  2727

Observation:
Harlequin tops the top 10 publishers list with 7535 publications

#### Top 10 book Authors

In [16]:
books.groupby('Book-Author').size().reset_index().sort_values(by=0, ascending=False)[:10]

Book-Author    0
1088        Agatha Christie  632
100923  William Shakespeare  567
91188          Stephen King  524
4567          Ann M. Martin  423
13497         Carolyn Keene  373
30559       Francine Pascal  372
40000          Isaac Asimov  330
73342          Nora Roberts  315
7311       Barbara Cartland  307
14422       Charles Dickens  302

Observation:
Agatha Christie tops the top 10 author's list with 632 authored books

In [17]:
user_ratings = users.merge(ratings, on='User-ID')

In [18]:
user_ratings.head()

User-ID                   Location   Age        ISBN  Book-Rating
0        2  stockton, california, usa  18.0  0195153448            0
1        7        washington, dc, usa   NaN   034542252            0
2        8   timmins, ontario, canada   NaN  0002005018            5
3        8   timmins, ontario, canada   NaN  0060973129            0
4        8   timmins, ontario, canada   NaN  0374157065            0

In [19]:
user_ratings['Book-Rating'].unique()

array([ 0,  5,  6,  7, 10,  9,  3,  8,  4,  2,  1], dtype=int64)

In [20]:
ratings.groupby('Book-Rating').size().reset_index().sort_values(by=0, ascending=False)

Book-Rating       0
0             0  716109
8             8  103736
10           10   78610
7             7   76457
9             9   67541
5             5   50974
6             6   36924
4             4    8904
3             3    5996
2             2    2759
1             1    1770

Observation:
Most given rating is 0

In [21]:
user_ratings.groupby('User-ID').size().reset_index().sort_values(by=0, ascending=False)[:10]

User-ID      0
4213      11676  13602
74815    198711   7550
58113    153662   6109
37356     98391   5891
13576     35859   5850
80185    212898   4785
105111   278418   4533
28884     76352   3367
42037    110973   3100
88584    235105   3067

Observation:
User with id 4213 has given the most number of ratings (13602) folllowed by user with id 74815 with 7550 ratings

In [22]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [23]:
books['Year-Of-Publication'].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

In [24]:
books.groupby('Year-Of-Publication').size().reset_index().sort_values(by=0, ascending=False)[:20]

Year-Of-Publication      0
93                 2002  13903
92                 2001  13715
90                 1999  13414
91                 2000  13373
89                 1998  12116
94                 2003  11610
88                 1997  11494
87                 1996  10687
86                 1995  10259
85                 1994   8857
84                 1993   7920
83                 1992   7390
82                 1991   6926
81                 1990   6394
80                 1989   5825
79                 1988   5545
78                 1987   4761
95                 2004   4629
77                 1986   4258
185                1999   4017

Observation:
Majority of books cam during the period 1994-2003

In [25]:
book_ratings=books.merge(ratings, on='ISBN')

In [26]:
book_ratings.head()

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   
2  0002005018         Clara Callan  Richard Bruce Wright                2001   
3  0002005018         Clara Callan  Richard Bruce Wright                2001   
4  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5  
2  http://images.amazon.com/images/P/0002005018.0...    11400            0  
3  http://images.amazon.com/images/P/0002005018.0...    11676            8  
4  http://images.amazon.com/images/P/0002005018.0...    41385            0

In [27]:
df=book_ratings.groupby(['Book-Title','Book-Rating']).size().reset_index().sort_values(by=0, ascending=False)
df

Book-Title  Book-Rating     0
396551                                      Wild Animus            0  1921
330322                        The Lovely Bones: A Novel            0   588
9111                                    A Painted House            0   472
47393                             Bridget Jones's Diary            0   438
335565                       The Nanny Diaries: A Novel            0   435
...                                                 ...          ...   ...
161205                                 Jingle Bell Rock           10     1
161204                                 Jingle Bell Rock            8     1
161203                                 Jingle Bell Rock            7     1
161199  Jingle Bell Bride (Silhouette Romance, No 1123)            0     1
406354                                Ã?Â?thique en toc            8     1

[406355 rows x 3 columns]

In [28]:
df[df['Book-Rating']==0][:10]

Book-Title  Book-Rating     0
396551                                      Wild Animus            0  1921
330322                        The Lovely Bones: A Novel            0   588
9111                                    A Painted House            0   472
47393                             Bridget Jones's Diary            0   438
335565                       The Nanny Diaries: A Novel            0   435
89067   Divine Secrets of the Ya-Ya Sisterhood: A Novel            0   417
306959                                The Da Vinci Code            0   404
271021                           Snow Falling on Cedars            0   387
339690                                The Pelican Brief            0   383
348503                          The Secret Life of Bees            0   368

Observation:
Book titled Wild animus has received the most number of 0 ratings 

In [29]:
df[df['Book-Rating']==10][:10]

Book-Title  Book-Rating    0
306969                                  The Da Vinci Code           10  160
134719  Harry Potter and the Sorcerer's Stone (Harry P...           10  152
134684  Harry Potter and the Prisoner of Azkaban (Book 3)           10  150
330332                          The Lovely Bones: A Novel           10  148
134622   Harry Potter and the Chamber of Secrets (Book 2)           10  139
134644       Harry Potter and the Goblet of Fire (Book 4)           10  136
371226                              To Kill a Mockingbird           10  133
134659  Harry Potter and the Order of the Phoenix (Boo...           10  115
348512                            The Secret Life of Bees           10  112
313956  The Fellowship of the Ring (The Lord of the Ri...           10  103

Observation:
However rating 10 is evenly distributed,
Book titled The Da Vinci Code has received the most number of 0 ratings

In [30]:
df[df['Book-Title']=='Wild Animus'][:10]

Book-Title  Book-Rating     0
396551  Wild Animus            0  1921
396556  Wild Animus            5    88
396554  Wild Animus            3    87
396553  Wild Animus            2    82
396552  Wild Animus            1    74
396557  Wild Animus            6    66
396558  Wild Animus            7    60
396555  Wild Animus            4    58
396559  Wild Animus            8    40
396560  Wild Animus            9    16

In [31]:
df2=book_ratings.groupby(['Book-Author','Book-Rating']).size().reset_index().sort_values(by=0, ascending=False)
df2

Book-Author  Book-Rating     0
143257         Nora Roberts            0  5491
177859         Stephen King            0  5414
95062          John Grisham            0  3460
83038       James Patterson            0  3458
129698   Mary Higgins Clark            0  3100
...                     ...          ...   ...
102234  Julie Fisher Gunter            7     1
102233  Julie Fisher Gunter            0     1
102232         Julie Firman            0     1
102231         Julie Ferris            0     1
99459        Joseph W. Rock            0     1

[198918 rows x 3 columns]

In [32]:
df2[df2['Book-Rating']==0][:10]

Book-Author  Book-Rating     0
143257        Nora Roberts            0  5491
177859        Stephen King            0  5414
95062         John Grisham            0  3460
83038      James Patterson            0  3458
129698  Mary Higgins Clark            0  3100
42434       Dean R. Koontz            0  2838
187975          Tom Clancy            0  2813
38189       Danielle Steel            0  2650
179874         Sue Grafton            0  2222
154966         R. L. Stine            0  2066

Observation: Books of Nora Roberts has received the most number of 0 ratings, followed by Stephen King

In [33]:
df2[df2['Book-Rating']==10][:10]

Book-Author  Book-Rating     0
177869         Stephen King           10  1051
78092         J. K. Rowling           10   885
143267         Nora Roberts           10   588
83048       James Patterson           10   408
95072          John Grisham           10   334
196916  William Shakespeare           10   313
78394      J. R. R. Tolkien           10   281
142040      Nicholas Sparks           10   274
37050             Dan Brown           10   266
85155       Janet Evanovich           10   263

Observation: Books of Stephen King has received the most number of 10 ratings, followed by JK rowling 

In [34]:
df2[df2['Book-Author']=='Stephen King'][:10]

Book-Author  Book-Rating     0
177859  Stephen King            0  5414
177869  Stephen King           10  1051
177867  Stephen King            8  1026
177868  Stephen King            9   828
177866  Stephen King            7   694
177864  Stephen King            5   565
177865  Stephen King            6   311
177863  Stephen King            4    89
177862  Stephen King            3    38
177861  Stephen King            2    28

# Popularity-based Recommender System

In [35]:
ratings_with_names = ratings.merge(books, on='ISBN')
ratings_with_names

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1         276726  0155061224            5   
2         276727  0446520802            0   
3         276729  052165615X            3   
4         276729  0521795028            6   
...          ...         ...          ...   
1031131   276704  0876044011            0   
1031132   276704  1563526298            9   
1031133   276706  0679447156            0   
1031134   276709  0515107662           10   
1031135   276721  0590442449           10   

                                                Book-Title        Book-Author  \
0                                     Flesh Tones: A Novel         M. J. Rose   
1                                         Rites of Passage         Judith Rae   
2                                             The Notebook    Nicholas Sparks   
3                                           Help!: Level 1      Philip Prowse   
4        The Amsterdam Connection : Level 4 (Cambridge ...        Sue Leather   
...                                                    ...                ...   
1031131  Edgar Cayce on the Akashic Records: The Book o...  Kevin J. Todeschi   
1031132  Get Clark Smart : The Ultimate Guide for the S...       Clark Howard   
1031133  Eight Weeks to Optimum Health: A Proven Progra...        Andrew Weil   
1031134   The Sherbrooke Bride (Bride Trilogy (Paperback))  Catherine Coulter   
1031135                                  Fourth Grade Rats     Jerry Spinelli   

        Year-Of-Publication  \
0                      2002   
1                      2001   
2                      1996   
3                      1999   
4                      2001   
...                     ...   
1031131                1998   
1031132                2000   
1031133                1997   
1031134                1996   
1031135                1996   

                                                 Publisher  \
0                                         Ballantine Books   
1                                                   Heinle   
2                                             Warner Books   
3                               Cambridge University Press   
4                               Cambridge University Press   
...                                                    ...   
1031131  A.R.E. Press (Association of Research &amp; Enlig   
1031132                                   Longstreet Press   
1031133                                    Alfred A. Knopf   
1031134                                         Jove Books   
1031135                                         Scholastic   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  http://images.amazon.com/images/P/0515107662.0...   
1031135  http://images.amazon.com/images/P/0590442449.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  htt

In [36]:
num_rating_df = ratings_with_names.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating': 'num_ratings'}, inplace=True)
num_rating_df

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [37]:
avg_rating_df = ratings_with_names.groupby('Book-Title')['Book-Rating'].agg(lambda x: x.astype(float).mean()).reset_index()
avg_rating_df.rename(columns={'Book-Rating': 'avg_ratings'}, inplace=True)
avg_rating_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [38]:
avg_rating_df.sort_values(by='avg_ratings', ascending=False)[140:160]

Book-Title  avg_ratings
39083                                   Constance : Poems         10.0
72660                       Garden Patterns &amp; Mosaics         10.0
212669                    The Wall (Reading Rainbow Book)         10.0
39070                                        Conspiracy X         10.0
163049                                   Socks for Supper         10.0
14268   Angel Reborn (Battle Angel Alita: Last Order, ...         10.0
163050  Socks, Socks, Socks: 70 Winning Patterns from ...         10.0
163056  Socrates, Buddha, Confucius, Jesus: From The G...         10.0
39057                            Consolation For An Exile         10.0
72647   Garden Details: Accents, Ornaments, and Finish...         10.0
122141            Moments: The Pulitzer Prize Photographs         10.0
72641   Garden Antiques and Collectibles (For Your Gar...         10.0
96276                                         Jesus Freak         10.0
72639                              Garcia Lorca Enamorado         10.0
72638                           Garcia : An American Life         10.0
14334                   Angel's Devil (A Regency Romance)         10.0
122139  Moments of being: Unpublished autobiographical...         10.0
72637                                              Garcia         10.0
163078  Soft Coated Wheaten Terriers: Everything About...         10.0
96170                        Jerusalem: Rebirth of a city         10.0

In [39]:
popular_df = num_rating_df.merge(avg_rating_df, on='Book-Title')

In [40]:
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [41]:
popular_df =  popular_df[popular_df['num_ratings']>=250].sort_values(by='avg_ratings', ascending = False)[:50]
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233851       White Oleander : A Novel (Oprah's Book Club)          356   

        avg_ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
80414      5.183453  
191612     5.007117  
187377     4.948370  
80445      4.895652  
211384     4.880769  
219741     4.700000  
183573     4.642539  
187880  

In [42]:
popular_df = popular_df.merge(books, on='Book-Title').drop_duplicates('Book-Title')[['Book-Title', 'Book-Author', 'avg_ratings', 'num_ratings', 'Image-URL-M']]

In [43]:
popular_df.shape

(50, 5)

# Collaborative Filtering Based Recommender system

In [44]:
x = ratings_with_names.groupby('User-ID').count()['Book-Rating']>200
good_users = x[x].index
good_users

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

In [45]:
filtered_rating = ratings_with_names[ratings_with_names['User-ID'].isin(good_users)]
filtered_rating.head()

User-ID        ISBN  Book-Rating  \
1150   277427  002542730X           10   
1151   277427  0026217457            0   
1152   277427  003008685X            8   
1153   277427  0030615321            0   
1154   277427  0060002050            0   

                                             Book-Title  \
1150  Politically Correct Bedtime Stories: Modern Ta...   
1151                 Vegetarian Times Complete Cookbook   
1152                                           Pioneers   
1153   Ask for May, Settle for June (A Doonesbury book)   
1154                  On a Wicked Dawn (Cynster Novels)   

                Book-Author Year-Of-Publication                  Publisher  \
1150      James Finn Garner                1994  John Wiley &amp; Sons Inc   
1151             Lucy  Moll                1995      John Wiley &amp; Sons   
1152  James Fenimore Cooper                1974           Thomson Learning   
1153          G. B. Trudeau                1982        Henry Holt &amp; Co   
1154      Stephanie Laurens                2002                 Avon Books   

                                            Image-URL-S  \
1150  http://images.amazon.com/images/P/002542730X.0...   
1151  http://images.amazon.com/images/P/0026217457.0...   
1152  http://images.amazon.com/images/P/003008685X.0...   
1153  http://images.amazon.com/images/P/0030615321.0...   
1154  http://images.amazon.com/images/P/0060002050.0...   

                                            Image-URL-M  \
1150  http://images.amazon.com/images/P/002542730X.0...   
1151  http://images.amazon.com/images/P/0026217457.0...   
1152  http://images.amazon.com/images/P/003008685X.0...   
1153  http://images.amazon.com/images/P/0030615321.0...   
1154  http://images.amazon.com/images/P/0060002050.0...   

                                            Image-URL-L  
1150  http://images.amazon.com/images/P/002542730X.0...  
1151  http://images.amazon.com/images/P/0026217457.0...  
1152  http://images.amazon.com/images/P/003008685X.0...  
1153  http://images.amazon.com/images/P/0030615321.0...  
1154  http://images.amazon.com/images/P/0060002050.0...

In [46]:
y = filtered_rating.groupby('Book-Title').count()['Book-Rating'] >= 50
popular_books = y[y].index
popular_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [47]:
final_df = filtered_rating[filtered_rating['Book-Title'].isin(popular_books)]
final_df.head()

User-ID        ISBN  Book-Rating  \
1150   277427  002542730X           10   
1163   277427  0060930535            0   
1165   277427  0060934417            0   
1168   277427  0061009059            9   
1174   277427  006440188X            0   

                                             Book-Title  \
1150  Politically Correct Bedtime Stories: Modern Ta...   
1163                      The Poisonwood Bible: A Novel   
1165                                 Bel Canto: A Novel   
1168  One for the Money (Stephanie Plum Novels (Pape...   
1174                                  The Secret Garden   

                  Book-Author Year-Of-Publication                  Publisher  \
1150        James Finn Garner                1994  John Wiley &amp; Sons Inc   
1163       Barbara Kingsolver                1999                  Perennial   
1165             Ann Patchett                2002                  Perennial   
1168          Janet Evanovich                1995                HarperTorch   
1174  Frances Hodgson Burnett                1998               HarperTrophy   

                                            Image-URL-S  \
1150  http://images.amazon.com/images/P/002542730X.0...   
1163  http://images.amazon.com/images/P/0060930535.0...   
1165  http://images.amazon.com/images/P/0060934417.0...   
1168  http://images.amazon.com/images/P/0061009059.0...   
1174  http://images.amazon.com/images/P/006440188X.0...   

                                            Image-URL-M  \
1150  http://images.amazon.com/images/P/002542730X.0...   
1163  http://images.amazon.com/images/P/0060930535.0...   
1165  http://images.amazon.com/images/P/0060934417.0...   
1168  http://images.amazon.com/images/P/0061009059.0...   
1174  http://images.amazon.com/images/P/006440188X.0...   

                                            Image-URL-L  
1150  http://images.amazon.com/images/P/002542730X.0...  
1163  http://images.amazon.com/images/P/0060930535.0...  
1165  http://images.amazon.com/images/P/0060934417.0...  
1168  http://images.amazon.com/images/P/0061009059.0...  
1174  http://images.amazon.com/images/P/006440188X.0...

In [48]:
final_df.shape

(58586, 10)

In [49]:
pt = final_df.pivot_table(index = 'Book-Title', columns = 'User-ID', values = 'Book-Rating')
pt.fillna(0, inplace = True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [50]:
from sklearn.metrics.pairwise import cosine_similarity

In [51]:
similarity_score = cosine_similarity(pt)
similarity_score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [52]:
def recommend(book_name):
    index =  np.where(pt.index == book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_score[index])), key=lambda x:x[1], reverse=True)[1:6]
    
    data=[]
    for i in similar_items:
        item=[]
        temp_df = books[books['Book-Title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    return data

In [53]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [54]:
sorted(list(enumerate(similarity_score[120])), key=lambda x:x[1], reverse=True)[1:6]

[(190, 0.4203141746966238),
 (664, 0.33617670904670893),
 (186, 0.3106624600581585),
 (340, 0.2977381938532878),
 (208, 0.28639015757618047)]

In [55]:
books.dropna(inplace=True)

In [56]:
recommend('The Secret Garden')

[["Jewel (Oprah's Book Club)",
  'Bret Lott',
  'http://images.amazon.com/images/P/0671042572.01.MZZZZZZZ.jpg'],
 ['That Camden Summer',
  'Lavyrle Spencer',
  'http://images.amazon.com/images/P/0399141200.01.MZZZZZZZ.jpg'],
 ['A Walk to Remember',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446608955.01.MZZZZZZZ.jpg'],
 ['GARDEN OF SHADOWS (Dollanger Saga (Paperback))',
  'Ariel Jennifer Jones',
  'http://images.amazon.com/images/P/067164257X.01.MZZZZZZZ.jpg'],
 ['Circle of Friends',
  'Maeve Binchy',
  'http://images.amazon.com/images/P/0440211263.01.MZZZZZZZ.jpg']]

In [57]:
import pickle
pickle.dump(popular_df, open('popular.pkl', 'wb'))

In [58]:
pickle.dump(pt,open('pt.pkl','wb'))
pickle.dump(books,open('books.pkl','wb'))
pickle.dump(similarity_score,open('similarity_score.pkl','wb'))

In [59]:
len(list(books['Book-Title'].unique()))

242129

In [60]:
books.shape

(271353, 8)

In [61]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    